File: 47.permutations-ii.go
package leetcode

import "slices"

// @leet start
func permuteUnique(nums []int) [][]int {
	n := len(nums)
	ans := [][]int{}
	curr := []int{}
	visited := make([]bool, n)
	// sort for the line below
	slices.Sort(nums)

	var backtrack func()
	backtrack = func() {
		if len(curr) == len(nums) {
			copied := make([]int, len(curr))
			copy(copied, curr)
			ans = append(ans, copied)
			return
		}

		for i := range nums {
			if visited[i] {
				continue
			}
			// this is the key line difference here to ensure prev elements are visited
			// before the current element if they're the same.
			// This ensures that the prev one can be permuted
			if i > 0 && nums[i-1] == nums[i] && !visited[i-1] {
				continue
			}

			visited[i] = true
			curr = append(curr, nums[i])
			backtrack()
			curr = curr[:len(curr)-1]
			visited[i] = false
		}

	}
	backtrack()

	return ans

}

// @leet end

File: 47.permutations-ii.py
from typing import List


# @leet start
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        ans = []
        n = len(nums)
        nums.sort()
        visited = [False] * n

        def dfs(curr: list[int], visited: list[bool]) -> None:
            if len(curr) == n:
                ans.append(curr[:])
            for i in range(n):
                if i > 0 and nums[i] == nums[i - 1] and not visited[i - 1]:
                    continue
                if visited[i]:
                    continue
                curr.append(nums[i])
                visited[i] = True
                dfs(curr, visited)
                curr.pop()
                visited[i] = False

        dfs([], visited)
        return ans


# @leet end

x = 0
nums = [1, 1, 2]
x = Solution().permuteUnique(nums)
print(x)

nums = [1, 2, 3]
x = Solution().permuteUnique(nums)
print(x)

